### PR TITLE
daemon/c8d: Implement import

### PR DESCRIPTION
**Pull request by vvoland**
**- What I did**
Implemented `docker import` for containerd integration

**- How I did it**
Write the blob to the content store, calculate the uncompressed data digest and create a manifest and image config for the selected platform (default is host platform if none provided).
I also extracted the common logic (parsing image name into reference and getting a data stream) to the API routes.

If the imported layer archive is uncompressed, it gets compressed with gzip before writing to the content store.
Archives compressed with gzip and zstd are imported as-is.
Xz and bzip2 are recompressed into gzip.

**- How to verify it**

1. Export some image from non-c8d Docker
```bash
$ docker run --name asdf ubuntu sh -c 'apt-get -y update && apt-get install -y fish'
$ docker export asdf >asdf.tar
```

2. Compress it with supported compressions (gz, zstd, xz, bzip2)
3. Import and verify with ctr (docker run also works, but is not yet upstreamed)
<details>
<summary>raw</summary>

```bash
$ docker import asdf.tar asdf:raw
sha256:5cebbe53a996ec2ad6c4101143a8c13f8d6999e5144a75ed21570fb96c93bd6b

# Manifest
$ ctr content get "sha256:5cebbe53a996ec2ad6c4101143a8c13f8d6999e5144a75ed21570fb96c93bd6b"
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "config": {
    "mediaType": "application/vnd.oci.image.config.v1+json",
    "digest": "sha256:1150939d6b988df2a1febb4b0fcbdc5245eb4c8335061b2ee239f878f0530a8e",
    "size": 291
  },
  "layers": [
    {
      "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
      "digest": "sha256:0953a62d60750e4932d6dbd7f79b9cf41c371f768ab4a7c7d384c1cdc75ffc99",
      "size": 9934617
    }
  ]
}


# Config
$ ctr content get "sha256:1150939d6b988df2a1febb4b0fcbdc5245eb4c8335061b2ee239f878f0530a8e"
{
  "created": "2022-12-16T14:54:19.593382044Z",
  "architecture": "arm64",
  "os": "linux",
  "config": {},
  "rootfs": {
    "type": "layers",
    "diff_ids": [
      "sha256:4aa1b69e24b6c5dfde91e20440f2feb0c2e85ec0596adef1bd1107005268a826"
    ]
  },
  "history": [
    {
      "created": "2022-12-16T14:54:19.593382044Z",
      "comment": "Imported from -"
    }
  ]
}


# Layer
$ ctr content get "sha256:0953a62d60750e4932d6dbd7f79b9cf41c371f768ab4a7c7d384c1cdc75ffc99" | gzip -d | tar --list | grep 'bin/fish'
usr/bin/fish
usr/bin/fish_indent
usr/bin/fish_key_reader

```
</details>

<details>
<summary>gz</summary>

```bash
$ docker import asdf.tar.gz asdf:gz
sha256:c2b5ae5a12eac4c397445289c1864c5d658b0e958ff1cd83276ccb48d32edadb

# Manifest
$ ctr content get "sha256:c2b5ae5a12eac4c397445289c1864c5d658b0e958ff1cd83276ccb48d32edadb"
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "config": {
    "mediaType": "application/vnd.oci.image.config.v1+json",
    "digest": "sha256:892ea7918be6bf49d44b7ef410d4c4eb5212afad5aa101ef563540f51442521c",
    "size": 291
  },
  "layers": [
    {
      "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
      "digest": "sha256:ad31c4743384e7d6be2518fd5d411d4175566057a7e38f6018f577d3132a4d71",
      "size": 9695278
    }
  ]
}


# Config
$ ctr content get "sha256:892ea7918be6bf49d44b7ef410d4c4eb5212afad5aa101ef563540f51442521c"
{
  "created": "2022-12-16T14:54:20.091913877Z",
  "architecture": "arm64",
  "os": "linux",
  "config": {},
  "rootfs": {
    "type": "layers",
    "diff_ids": [
      "sha256:4aa1b69e24b6c5dfde91e20440f2feb0c2e85ec0596adef1bd1107005268a826"
    ]
  },
  "history": [
    {
      "created": "2022-12-16T14:54:20.091913877Z",
      "comment": "Imported from -"
    }
  ]
}


# Layer
$ ctr content get "sha256:ad31c4743384e7d6be2518fd5d411d4175566057a7e38f6018f577d3132a4d71" | gzip -d | tar --list | grep 'bin/fish'
usr/bin/fish
usr/bin/fish_indent
usr/bin/fish_key_reader

```
</details>

<details>
<summary>zstd</summary>

```bash
$ docker import asdf.tar.zst asdf:zstd
sha256:a49f3a56f566e6a9741b5b0a568768c007cb4a962c506cbb2867d283384795d0

# Manifest
$ ctr content get "sha256:a49f3a56f566e6a9741b5b0a568768c007cb4a962c506cbb2867d283384795d0"
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "config": {
    "mediaType": "application/vnd.oci.image.config.v1+json",
    "digest": "sha256:f5357c0fb73a3681c1ae09749df39e2b1d3e379ca2c92498a3aafe020414e413",
    "size": 291
  },
  "layers": [
    {
      "mediaType": "application/vnd.oci.image.layer.v1.tar+zstd",
      "digest": "sha256:7f7585dbea4e0ea7aa7790d4b81f0c324fe801e77e28a372e46e4c96adda3fc0",
      "size": 9513145
    }
  ]
}


# Config
$ ctr content get "sha256:f5357c0fb73a3681c1ae09749df39e2b1d3e379ca2c92498a3aafe020414e413"
{
  "created": "2022-12-16T14:54:20.491820044Z",
  "architecture": "arm64",
  "os": "linux",
  "config": {},
  "rootfs": {
    "type": "layers",
    "diff_ids": [
      "sha256:4aa1b69e24b6c5dfde91e20440f2feb0c2e85ec0596adef1bd1107005268a826"
    ]
  },
  "history": [
    {
      "created": "2022-12-16T14:54:20.491820044Z",
      "comment": "Imported from -"
    }
  ]
}


# Layer
$ ctr content get "sha256:7f7585dbea4e0ea7aa7790d4b81f0c324fe801e77e28a372e46e4c96adda3fc0" | zstd -d | tar --list | grep 'bin/fish'
usr/bin/fish
usr/bin/fish_indent
usr/bin/fish_key_reader

```
</details>

<details>
<summary>xz</summary>

```bash
$ docker import asdf.tar.xz asdf:xz
sha256:0c867e1189a713d170c8afec8f75b9dc67f204d9e49c95b8626a99a848cd681a

# Manifest
$ ctr content get "sha256:0c867e1189a713d170c8afec8f75b9dc67f204d9e49c95b8626a99a848cd681a"
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "config": {
    "mediaType": "application/vnd.oci.image.config.v1+json",
    "digest": "sha256:de62b7c145c11db01ddebfe4373a224c462c7885fa9b5006dde141c18ba157f6",
    "size": 291
  },
  "layers": [
    {
      "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
      "digest": "sha256:0953a62d60750e4932d6dbd7f79b9cf41c371f768ab4a7c7d384c1cdc75ffc99",
      "size": 9934617
    }
  ]
}


# Config
$ ctr content get "sha256:de62b7c145c11db01ddebfe4373a224c462c7885fa9b5006dde141c18ba157f6"
{
  "created": "2022-12-16T14:54:21.443434628Z",
  "architecture": "arm64",
  "os": "linux",
  "config": {},
  "rootfs": {
    "type": "layers",
    "diff_ids": [
      "sha256:4aa1b69e24b6c5dfde91e20440f2feb0c2e85ec0596adef1bd1107005268a826"
    ]
  },
  "history": [
    {
      "created": "2022-12-16T14:54:21.443434628Z",
      "comment": "Imported from -"
    }
  ]
}


# Layer
$ ctr content get "sha256:0953a62d60750e4932d6dbd7f79b9cf41c371f768ab4a7c7d384c1cdc75ffc99" | gzip -d | tar --list | grep 'bin/fish'
usr/bin/fish
usr/bin/fish_indent
usr/bin/fish_key_reader

```
</details>

<details>
<summary>bz2</summary>

```bash
$ docker import asdf.tar.bz2 asdf:bz2
sha256:d52042ccf7a9b929544e8511d76f02d58f0162dd9acb1d837ad1717dd2e27dde

# Manifest
$ ctr content get "sha256:d52042ccf7a9b929544e8511d76f02d58f0162dd9acb1d837ad1717dd2e27dde"
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "config": {
    "mediaType": "application/vnd.oci.image.config.v1+json",
    "digest": "sha256:47bfa38ef6af5ca2e4c8b0460d367fcce5430443f7ce22400bc3ad233e415014",
    "size": 291
  },
  "layers": [
    {
      "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
      "digest": "sha256:0953a62d60750e4932d6dbd7f79b9cf41c371f768ab4a7c7d384c1cdc75ffc99",
      "size": 9934617
    }
  ]
}


# Config
$ ctr content get "sha256:47bfa38ef6af5ca2e4c8b0460d367fcce5430443f7ce22400bc3ad233e415014"
{
  "created": "2022-12-16T14:54:23.271932879Z",
  "architecture": "arm64",
  "os": "linux",
  "config": {},
  "rootfs": {
    "type": "layers",
    "diff_ids": [
      "sha256:4aa1b69e24b6c5dfde91e20440f2feb0c2e85ec0596adef1bd1107005268a826"
    ]
  },
  "history": [
    {
      "created": "2022-12-16T14:54:23.271932879Z",
      "comment": "Imported from -"
    }
  ]
}


# Layer
$ ctr content get "sha256:0953a62d60750e4932d6dbd7f79b9cf41c371f768ab4a7c7d384c1cdc75ffc99" | gzip -d | tar --list | grep 'bin/fish'
usr/bin/fish
usr/bin/fish_indent
usr/bin/fish_key_reader

```
</details>

**- Description for the changelog**
Support `docker import` when containerd snapshotter is enabled

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/containerd/image_import.go
Comment (vvoland) [inline daemon/containerd/image_import.go:149]: Will uncomment once related changes are upstreamed.

File: daemon/containerd/image_import.go
Comment (cpuguy83) [reply daemon/containerd/image_import.go:175]: Then the archive won't be very portable.

**Review by corhere — CHANGES_REQUESTED** [review]

File: daemon/images/image_import.go
Comment (vvoland) [reply daemon/images/image_import.go:25]: Totally forgot about this! Thanks for the reminder!

File: api/server/router/image/backend.go
Comment (vvoland) [reply api/server/router/image/backend.go:35]: Thanks!

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go:47]: Thanks, I added the platform check in wrong place 🤦🏻

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go:67]: Good catch, thanks!
I used logger here to disambiguate from the standard library `log`.

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go:85]: Thanks, fixed.

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go:281]: Thanks, fixed

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go:58]: Ah, I missed this because I copy-pasted this bit from archive.DecompressStream, which does the Put in other place 😄 
Thanks for catching this!

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go:281]: Done.

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go:173]: Thanks, fixed.
(too much C/C++ in my past)

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go:237]: I split it into 2 functions.
Currently we unpack all images as soon as we get them and all image-related operations assume that they are unpacked.

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go:254]: Done, thanks.

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go:161]: I don't think we can use t he TeeReader trick here, because decompressed stream is different Reader than the uncompressed one. 
I implemented it with another `io.Pipe` (meh), but maybe there's some clever way to work around this?

File: daemon/containerd/image_import.go
Comment (corhere) [reply daemon/containerd/image_import.go]: > all image-related operations assume that they are unpacked.

Could that assumption be violated by images which were pulled by `ctr`, `nerdctl` or another containerd client?

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go]: I wrote the pipe error handling a little differently now.
Also, with `CloseWithError`, I think handling the ctx.Done separately is not needed.
If the ctx is cancelled, then `CloseWithError` will set it on the pipe, and then io.Copy will return with that err.

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go]: Yeah, if other client does something our namespace (which is "moby", not the default) then it definitely can break this.
For now we assume that if user chooses to fiddle with moby namespace then he knows what he is doing.

File: daemon/images/image_import.go
Comment (corhere) [inline daemon/images/image_import.go:23]: That's not quite true: the uncompressed layer archive is passed to the graphdriver, not recompressed. And as far as I am aware, none of our graph drivers store compressed archives of layers anywhere.

File: daemon/containerd/image_import.go
Comment (corhere) [inline daemon/containerd/image_import.go:66]: I got curious on Friday when working on my own stuff and peeked under the hood at [how `(*logrus.Entry).WithField` is implemented.](https://github.com/sirupsen/logrus/blob/f8bf7650dccb756cea26edaf9217aab85500fe07/entry.go#L119-L122) 
```go
return entry.WithFields(Fields{key: value})
```
Turns out it's both more verbose and costs more allocations to chain two `logger.WithField()` than to use `logger.WithFields(logrus.Fields{...})`.
```suggestion
	logger = logger.WithFields(logrus.Fields{
		"compressedDigest":   compressedDigest,
		"uncompressedDigest": uncompressedDigest,
	})
```

File: daemon/containerd/image_import.go
Comment (corhere) [inline daemon/containerd/image_import.go:70]: It looks a lot like `fillUncompressedLabel` is attaching some metadata to an existing blob in the content store, not writing a blob as the debug log message states.

**Review by corhere — APPROVED** [review]
I have a couple of nitpicks and optional suggestions, though nothing impactful. The parts which matter all LGTM!

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go]: Thanks, TIL! 😄

File: daemon/images/image_import.go
Comment (vvoland) [reply daemon/images/image_import.go]: Thanks, fixed!

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go]: Thanks, fixed.

File: daemon/containerd/image_import.go
Comment (rumpl) [inline daemon/containerd/image_import.go:49]:
```suggestion
	ctx, release, err := i.client.WithLease(ctx)
	if err != nil {
		return "", errdefs.System(err)
	}
	defer release(ctx)
```

Comment (vvoland) [reply daemon/containerd/image_import.go]: Thanks, done!

File: api/server/router/image/image_routes.go
Comment (cpuguy83) [inline api/server/router/image/image_routes.go:105]: TIL we can import remote tars.

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go:33]: Wouldn't it only work for applying layers though? We would still have an image with a non-standard `application/vnd.oci.image.layer.v1.tar+xz`/`application/vnd.oci.image.layer.v1.tar+bz2` layer, which would be awkward to push/save.

File: daemon/containerd/image_import.go
Comment (vvoland) [reply daemon/containerd/image_import.go]: Oh, good catch, looks like I somehow missed lost this one 🤔. Fixed now, thanks!

File: daemon/containerd/image_import.go
Comment (cpuguy83) [reply daemon/containerd/image_import.go:33]: Good point there and this is pretty much inline with what happens today.

**Comment by vvoland** [comment]
CI failure is not relevant to this PR, flaky test: `TestDockerCLILogsSuite/TestLogsFollowGoroutinesWithStdout`, created an issue for tracking: https://github.com/moby/moby/issues/44747

**Review by cpuguy83 — APPROVED** [review]

File: api/server/router/image/image_routes.go
Comment (thaJeztah) [inline api/server/router/image/image_routes.go:82]: Not new (existing code did this as well), but wondering if this should look for `.Digested`; I recall I had a PR open for that in another area; https://github.com/moby/moby/pull/44510

File: api/server/router/image/image_routes.go
Comment (thaJeztah) [inline api/server/router/image/image_routes.go:98]: Also not new, but my eye dropped on this, and wondering if we should look at this (in light of the containerd integration); I know quite some users use `docker import` to transfer / load images into a daemon. Currently (pre containerd) this discards the (registry-)digest, so while the image is (mostly) the same, it's not the same as doing a `docker pull` of the same image.

I wonder how far we get with making them the same (with the containerd integration), and in that case, we should avoid mutating the imported image; the `Imported from` may not be very relevant, and if setting this message by default may affect the imported image (making it not the same as `docker pull`).

I can create a ticket for that for a follow-up.

File: api/server/router/image/image_routes.go
Comment (thaJeztah) [reply api/server/router/image/image_routes.go:105]: LOL. I completely forgot about that too 😂

File: api/server/router/image/image_routes.go
Comment (thaJeztah) [inline api/server/router/image/image_routes.go:127]: We really need to cleanup that jsonmessage package; it's awkward that this function expects an ID as first argument, and that we don't pass any (we could add another utility that's clearer on intent, and is specific to stream an (informational) "message"), although the whole JSONMessage response is "fuzzy" (it's part of the API, but we don't have proper documentation on what to expect etc.

File: daemon/images/image_import.go
Comment (thaJeztah) [inline daemon/images/image_import.go:84]: Was returning the `id` on a failure intentional here? Are there uses for still consuming the `id` even if this failed?

I see the current code calling this is discarding it;

```go
id, progressErr = ir.backend.ImportImage(ctx, ref, platform, comment, layerReader, r.Form["changes"])

if progressErr == nil {
	output.Write(streamformatter.FormatStatus("", id.String()))
}
```


_IF_ there's uses for this and we keep this return, we should add a comment here, and also as part of the function's documentation to explain in what cases both an ID, and an error are returned.

**Review by thaJeztah — COMMENTED** [review]
Overall LGTM, but left some comments for follow-ups, and a question about the `id` returned (that one should probably be addressed before merging).

File: api/server/router/image/image_routes.go
Comment (vvoland) [reply api/server/router/image/image_routes.go:98]: I'm +1 on removing this - IMO it doesn't add any value, especially when in most cases src is `-`.
I'll remove it and open a follow up PR so we can discuss this and not block this PR.

File: daemon/images/image_import.go
Comment (vvoland) [reply daemon/images/image_import.go:84]: From the perspective of this function, the image has been created and the ID exists, it's just the tag that failed. Since we don't remove the created image it makes some sense to me that we both return an id and error here.

But you're right that the current usage doesn't make any use of it. I'll change that, push a separate commit and we can see if it makes more sense.

Comment (thaJeztah) [reply daemon/images/image_import.go:84]: Yes, I had that scenario in mind indeed; effectively, we're somewhat pretending the `ImportImage()` to be an atomic operation, but it's not fully (image imported, but not tagged). In order to make the `id` still usable for that scenario, it should either return a specific error (so that the caller can handle it differently), or the operations should be split (`import -> tag`).

Mostly wondering what happens if we import successfully, but fail to tag?

- Is the image accessible (by digest?)
- If so, can it be (manually) tagged later (once / if we support `docker image tag <digest> name:tag`)
- Will it be garbage collected if no action takes place after this (and _when_)?

Comment (vvoland) [reply daemon/images/image_import.go:84]: Before containerd image service, if the image fails to be tagged then it's still accessible.
This is what I get when I just replace the call to `TagImageWithReference` with `err = errors.New("simulated error")`
```bash
root@37d238ca2d8a:/go/src/github.com/docker/docker# docker import asdf.tar.zst mynewimage
simulated error
sha256:85bfb97409716fd76c06a64ebfabc4de87a62e12b40b36d322543a093910de08
root@37d238ca2d8a:/go/src/github.com/docker/docker# docker image ls
REPOSITORY   TAG       IMAGE ID       CREATED         SIZE
<none>       <none>    85bfb9740971   6 seconds ago   22.3MB
root@37d238ca2d8a:/go/src/github.com/docker/docker# docker tag 85bfb9740971 mynewimage
root@37d238ca2d8a:/go/src/github.com/docker/docker# docker image ls
REPOSITORY   TAG       IMAGE ID       CREATED         SIZE
mynewimage   latest    85bfb9740971   6 minutes ago   22.3MB
```


With containerd it's not that easy, because tagging = creating an image, and if the image isn't created then there's nothing that references the underlying content so it will be GC'ed. In this case we should just error the whole operation with error.
That also makes it difficult to split the whole operation into import&tag - because the atomicity of import is dependent on the image service implementation.

File: api/server/router/image/image_routes.go
Comment (vvoland) [reply api/server/router/image/image_routes.go:82]: Yeah, good point, I recently did consider this in other context, but didn't remember that this one also does this. Fixed, thanks!

File: api/server/router/image/image_routes.go
Comment (vvoland) [reply api/server/router/image/image_routes.go:98]: https://github.com/moby/moby/pull/44801

**Comment by vvoland** [comment]
CI failure unrelated:
```
=== Failed
=== FAIL: github.com/docker/docker/integration/container TestHealthCheckProcessKilled (14.54s)
    main_test.go:32: assertion failed: error is not nil: Error response from daemon: Could not kill running container 00fb7c6182d3597e555afaf2b7e19c6f9562c062ca313220172601c855f440d0, cannot remove - tried to kill container, but did not receive an exit event: failed to remove 00fb7c6182d3597e555afaf2b7e19c6f9562c062ca313220172601c855f440d0

DONE 151 tests, 39 skipped, 1 failure in 226.429s
```

**Comment by neersighted** [comment]
It looks like a thought got cut off in the last commit message:

> This makes the import operation return both a valid image ID and an
error if the image was imported but the tag operation failed.
It makes it possible to tag

File: daemon/images/image_import.go
Comment (vvoland) [reply daemon/images/image_import.go:87]: I dropped this, see: https://github.com/moby/moby/pull/44628#issuecomment-1380046654

**Review by rumpl — APPROVED** [review]

**Review by neersighted — APPROVED** [review]

**Pull request merged**
